util/util.go

```diff
@@ -432,6 +432,9 @@ func GetGVKMetadata(ctx context.Context, c client.Client, gvk schema.GroupVersio
 // GetCRDWithContract retrieves a list of CustomResourceDefinitions from using controller-runtime Client,
 // filtering with the `contract` label passed in.
 // Returns the first CRD in the list that matches the GroupVersionKind, otherwise returns an error.
+// Deprecated: This func is going to be removed in a next release.
+// NOTE: This func is using a client.Client to list CRDs and thus implicitly creates an informer for CRDs
+// which leads to high memory consumption.
 func GetCRDWithContract(ctx context.Context, c client.Client, gvk schema.GroupVersionKind, contract string) (*apiextensionsv1.CustomResourceDefinition, error) {
```

Comment:
```suggestion
func GetCRDWithContract(ctx context.Context, c client.Reader, gvk schema.GroupVersionKind, contract string) (*apiextensionsv1.CustomResourceDefinition, error) {
```

Let's do a breaking change here? The new interface is a bit more strict which is a valuable thing to have, also going from `client.Client` to `client.Reader` is technically a compatible change, given that every Client is also a Reader

Reply:
Sounds good - done. 

WDYT is there a better way to handle the breaking behavioral change in the parent func too? (https://github.com/kubernetes-sigs/cluster-api/pull/5698#discussion_r752054269)